core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/score/constraint/DefaultIndictment.java

```diff
@@ -1,35 +1,48 @@
 package org.optaplanner.core.impl.score.constraint;
 
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import org.optaplanner.core.api.score.Score;
 import org.optaplanner.core.api.score.constraint.ConstraintMatch;
 import org.optaplanner.core.api.score.constraint.Indictment;
+import org.optaplanner.core.api.score.stream.ConstraintJustification;
 
 public final class DefaultIndictment<Score_ extends Score<Score_>> implements Indictment<Score_> {
 
-    private final Object justification;
-
-    private final Set<ConstraintMatch<Score_>> constraintMatchSet;
+    private final Object indictedObject;
+    private final Set<ConstraintMatch<Score_>> constraintMatchSet = new LinkedHashSet<>();
+    private List<ConstraintJustification> constraintJustificationList;
     private Score_ score;
 
-    public DefaultIndictment(Object justification, Score_ zeroScore) {
-        this.justification = justification;
-        constraintMatchSet = new LinkedHashSet<>();
-        score = zeroScore;
+    public DefaultIndictment(Object indictedObject, Score_ zeroScore) {
+        this.indictedObject = indictedObject;
+        this.score = zeroScore;
     }
 
     @Override
-    public Object getJustification() {
-        return justification;
+    public <IndictedObject_> IndictedObject_ getIndictedObject() {
+        return (IndictedObject_) indictedObject;
     }
 
     @Override
     public Set<ConstraintMatch<Score_>> getConstraintMatchSet() {
         return constraintMatchSet;
     }
 
+    @Override
+    public List<ConstraintJustification> getJustificationList() {
+        if (constraintJustificationList == null) {
+            constraintJustificationList = constraintMatchSet.stream()
+                    .map(s -> (ConstraintJustification) s.getJustification())
+                    .distinct()
+                    .collect(Collectors.toList());
+        }
+        return constraintJustificationList;
```

Comment:
This may return `null` if someone calls `addConstraintMatch()` or `removeConstraintMatch()` from a different thread at the same moment.

Reply:
True. That said, this class is not thread-safe (in more places than just this one), and I do not see a need for it to be.